File: Medium/1642. Furthest Building You Can Reach/solution.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        priority_queue<int> myQueue;
        int result = 0;
        for (int i = 1; i < heights.size(); i++){
            if (heights[i] <= heights[i - 1]){
                result++;
            }
            else{
                int need = heights[i] - heights[i - 1];
                if (need <= bricks){
                    bricks -= need;
                    myQueue.push(need);
                    result++;
                }
                else{
                    if (ladders == 0){
                        break;
                    }
                    if (myQueue.empty() || myQueue.top() <= need){
                        ladders--;
                    }
                    else{
                        ladders--;
                        bricks += myQueue.top();
                        myQueue.pop();
                        bricks -= need;
                        myQueue.push(need);
                    }
                    result++;
                }
            }
        }
        return result;
    }
};

File: Medium/1642. Furthest Building You Can Reach/solution.cs
public class Solution {
    public int FurthestBuilding(int[] heights, int bricks, int ladders) {
        PriorityQueue<int, int> myQueue = new PriorityQueue<int, int>();

        for (int i = 1; i < heights.Length; i++){
            int difference = heights[i] - heights[i - 1];

            if (difference > 0){
                bricks -= difference;
                myQueue.Enqueue(-difference, -difference);

                if (bricks < 0){
                    if (ladders == 0){
                        return i - 1;
                    }
                    else{
                        ladders--;
                        bricks -= myQueue.Dequeue();
                    }
                }
            }
        }

        return heights.Length - 1;
    }
}